### PR TITLE
Add Anaconda into the pre-built notebook pipeline

### PR DESCRIPTION
**Pull request by Xaenalt**
Draft of including the Anaconda notebook into pre-built pipeline

Related-to: https://github.com/opendatahub-io/notebooks/issues/110

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by atheo89** [comment]
Hi @Xaenalt, I checked out your PR and I would like to add some comments before you officially open it. 
First, I observed that the way that you approached building the anaconda notebook image is s2i, on this repo we don't use this way for building the notebooks, but on the contrary, we are doing it by giving all the instructions in a Dockerfile, as well as, each notebook inherits functionalities from the previous one. 
The second thing that I saw based on your requirements.in file is that you need almost the same packages as they are on the data science notebook. So, in that case, you can base your image from the data science notebook and add on that file only the packages that don't exist on the previous chain notebooks. 
So, in order to comply with this repo, you need to create a directory under `jupyter` folder with a name like for example `${NOTEBOOK_NAME}/ubi8-python-3.8`  with a Dockerfile that you install the Anaconda package manager and whatever else you need in order to be functional, (for your reference check the other notebooks based on the datascience image (PyTorch, TensorFlow)), the requirements.in file and the requirements.txt that will be generated after giving the correct  recipe into makefile that should be something like the following:
```
.PHONY: jupyter-${NOTEBOOK_NAME}-ubi8-python-3.8
jupyter-${NOTEBOOK_NAME}-ubi8-python-3.8: jupyter-datascience-ubi8-python-3.8
	$(call pip_compile,bootstrap/ubi8-python-3.8,\
		base/ubi8-python-3.8/requirements.in jupyter/minimal/ubi8-python-3.8/requirements.in jupyter/datascience/ubi8-python-3.8/requirements.in jupyter/${NOTEBOOK_NAME}/ubi8-python-3.8/requirements.in,\
		jupyter/${NOTEBOOK_NAME}/ubi8-python-3.8/requirements.txt)
	$(call image,$@,jupyter/${NOTEBOOK_NAME}/ubi8-python-3.8,$<)
```
Finally, test the changes locally, by running the `$ make jupyter-${NOTEBOOK_NAME}-ubi8-python-3.8` from the terminal

**Comment by atheo89** [comment]
I think you will not have any issues if you try to build your notebook from the [registry.access.redhat.com/ubi8/python-38:latest](https://catalog.redhat.com/software/containers/ubi8/python-38/5dde9cacbed8bd164a0af24a?container-tabs=dockerfile) instead of [registry.access.redhat.com/ubi8/s2i-base](https://catalog.redhat.com/software/containers/ubi8/s2i-base/5c83976a5a13464733ec6790?container-tabs=dockerfile)

**Comment by Xaenalt** [comment]
Right, I built a customized python-38 image, so if I have to use the original, it'll be a lot of uninstalling existing packages, and I had to patch the builder script to use Anaconda for requirements.in rather than pip. Anaconda essentially is a replacement for pip

**Comment by atheo89** [comment]
Hey @Xaenalt do you still need this PR?

**Comment by Xaenalt** [comment]
Ok, that should handle the base image, hopefully

**Comment by Xaenalt** [comment]
Ok, built and tested this locally, we should be all set!

**Comment by Xaenalt** [comment]
I'd like to do a bit of automation on updating the environment.yml stuff as packages are updated, similar workflow to the pipfile.lock stuff

**Comment by Xaenalt** [comment]
@atheo89 how does this look? Tested locally and it works, need to get it into the image push pipeline, and then I can update the dashboard bits to enable it in RHODS

File: jupyter/datascience/anaconda-python-3.8/requirements.in
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/requirements.in:2]: Use Pipefile instead requirements.in

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:13]: Mention the upstream repo here: `https://github.com/opendatahub-io/notebooks//s2i-minimal-notebook-anaconda`

**Review by atheo89 — CHANGES_REQUESTED** [review]

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:14]: Configure this registry for the downstream. 
For the upstream mirror the image under the `quay.io/repository/opendatahub/workbench-images`

**Review by atheo89 — CHANGES_REQUESTED** [review]

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/Dockerfile:13]: I thought this was going to be the main upstream repo instead

File: jupyter/datascience/anaconda-python-3.8/requirements.in
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/requirements.in:2]: This doesn't use pip, I just am using that to generate the exact versions of packages in the environment.yml

**Comment by Xaenalt** [comment]
https://github.com/openshift/release/pull/39863 for the ci pipeline

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:13]: This should be `https://github.com/opendatahub-io/notebooks/jupyter/datascience/anaconda-python-3.8`

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:4]: Is it Minimal or Datascience anacoda? I think it is a datascience!

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:11]: It needs to point `/opendatahub/workbench-images:jupyter-datascience-anaconda-python-3.8` registry

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:12]: main

**Comment by atheo89** [comment]
> [openshift/release#39863](https://github.com/openshift/release/pull/39863) for the ci pipeline

Added a comment, It needs a small fix

File: jupyter/datascience/anaconda-python-3.8/kustomize/base/statefulset.yaml
Comment (atheo89) [inline jupyter/datascience/anaconda-python-3.8/kustomize/base/statefulset.yaml:20]: Needs to be updated to the  quay.io/opendatahub/workbench-images

**Comment by Xaenalt** [comment]
> I tested the notebook locally and it looks good. But, I observed significant differences in the Elyra extensions between this notebook and the data science one. On the datascience we enable only the following: `elyra-pipeline-editor-extension`, `elyra-python-editor-extension`, `elyra-code-snippet-extension` , as shown in the image below: **Data Science Notebook** Screenshot from 2023-05-31 10-37-33 **Anacoda Data Science** Screenshot from 2023-05-31 11-11-13
> 
> I am not sure if we need all the extensions enabled in this anacoda notebook or not, or if it is due to misconfiguration. cc @harshad16

I kind of just installed the packages based on what was in the requirements.in from the regular datascience one. There wasn't a version of the s3 browser, but I tried to keep close to the datascience one. If I need to change stuff to remove other plugins I can

**Comment by Xaenalt** [comment]
Oh, I see, the ubi8-python-3.8 suffix is hardcoded... hmmm

**Comment by Xaenalt** [comment]
Ok, there we go, make deploy8-jupyter-datascience-anaconda-python-3.8 should now work as expected

**Review by harshad16 — CHANGES_REQUESTED** [review]
Thanks for the PR.

Noticed s2i script have been added in this PR.
Only assemble script has new changes as per this PR
rest of the s2i script are same as default
https://github.com/sclorg/s2i-python-container/tree/master/3.8/s2i/bin

can we remove those files (init-wrapper, usage, run) and 
only keep the one which has edit and can be included by this https://docs.openshift.com/container-platform/3.11/using_images/s2i_images/customizing_s2i_images.html

**Comment by Xaenalt** [comment]
@harshad16 it's a custom builder image, I don't think it'll have those in there if I don't add them in will it?

**Comment by harshad16** [comment]
> @harshad16 it's a custom builder image, I don't think it'll have those in there if I don't add them in will it?

In base-anacode image, it seem like we are using s2i image as base https://github.com/opendatahub-io/notebooks/blob/8f9343fa209e2160d0ec385f00bde410bb6a5c7d/base/anaconda-python-3.8/Dockerfile#L3
it will have s2i script with it.

**Comment by Xaenalt** [comment]
Ok cool, I wanted to provide that as a compatible base image in case we have other stuff that might need it

**Comment by harshad16** [comment]
> Ok cool, I wanted to provide that as a compatible base image in case we have other stuff that might need it

ack, my perspective to remove the additional s2i scripts, if we are not including new bits in it,
 is that we would need to maintain it as well, where as the upstream s2i-python is already doing that.
 If we are in for maintain these script, we can get this in.

**Comment by Xaenalt** [comment]
Right, makes sense, less maintenance on this the better

**Comment by Xaenalt** [comment]
So, any other changes we need on this?

**Comment by atheo89** [comment]
Hey @Xaenalt! What about the changes proposed by Harshad? 
Another thing is, once the PR on CI get merged (I added my review there) you can run on this PR `retest` to test the anaconda image.

**Comment by Xaenalt** [comment]
I removed the other files per Harshad's suggestion, and cool, hopefully the CI stuff will merge soon

**Comment by Xaenalt** [comment]
Hmmm, it's getting this error:

```
  * could not run steps: step jupyter-datascience-anaconda-python-3.8-pr-image-mirror failed: "jupyter-datascience-anaconda-python-3.8-pr-image-mirror" pre steps failed: failed to create or restart jupyter-datascience-anaconda-python-3.8-pr-image-mirror-opendatahub-io-ci-image-mirror pod: unable to create pod: Pod "jupyter-datascience-anaconda-python-3.8-pr-image-mirror-opendatahub-io-ci-image-mirror" is invalid: [spec.volumes[9].name: Invalid value: "jupyter-datascience-anaconda-python-3.8-pr-image-mirror": must not contain dots, spec.containers[0].volumeMounts[4].name: Not found: "jupyter-datascience-anaconda-python-3.8-pr-image-mirror"] 
```

Maybe a typo in my CI tests? I'm not sure where though, it looks pretty identical to the others, how does it get that volume name?

**Comment by Xaenalt** [comment]
@harshad16 any chance you can help me figure this one out?

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:36]: if elyra pipelines are being installed via the environment.yml 
please copy the changes from the data science notebook to keep the consistency.
https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi8-python-3.8/Dockerfile#L26-L44

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/environment.yml:60]: If this is just for the ODH community, then we are good with this
if this would be included in the RHODS,
please install each plugin separately so we have control.
https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi8-python-3.8/Pipfile#L10-L13

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/environment.yml:140]: currently 1.5.x version is only support by the Data science pipeline component of ODH.

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/environment.yml:131]: Seems like we are not including the `jupyterlab-s3-browser` plugin via environment.yml
is there some specific reason ?

File: jupyter/datascience/anaconda-python-3.8/setup-elyra.sh
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/setup-elyra.sh:1]: Please match this file https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi8-python-3.8/setup-elyra.sh
to keep consistency.

**Review by harshad16 — CHANGES_REQUESTED** [review]
Based on the cmd failure, we require this PR: https://github.com/openshift/release/pull/39979 for fixing the test.


Futher comments are based on plugins:
- we require to provide users with consistency across image, as this UBI8
based image it confuses me a little, if this is being provided already, then
we can remove elyra __pipeline__ plugins. 
if this is going to be provided as new, then please include the following suggestion.

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/environment.yml:60]: So, funny story, there isn't any more granular package, that's it. I can always just remove elyra stuff if we want

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/environment.yml:140]: Funny story again, there is no 1.5 version, https://anaconda.org/conda-forge/kfp/files

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/environment.yml:60]: Nvm, there is a more granular one, ok, I'll swap that

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/environment.yml:131]: No equivalent Anaconda package, there's a random third party one with 116 downloads, didn't feel like a good option

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/environment.yml:140]: Oh, if you meant kfp-tekton, there is a lower version, included kfp-tekton=1.5.1=pyhd8ed1ab_0

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (Xaenalt) [reply jupyter/datascience/anaconda-python-3.8/Dockerfile]: Only change from there is that there is no /opt/app-root/share/jupyter/labextensions/@elyra/theme-extension/static/lib_index_js.*.js files to change, so removed that line. Other than that updated to consistency with the ubi8 datascience one

**Comment by Xaenalt** [comment]
Blah, I messed up the package list, regenerating it

**Comment by Xaenalt** [comment]
Cool, ok now with the updated test, as soon as I get the updated pkg list we should be good to go hopefully

**Comment by atheo89** [comment]
Great, overall /lgtm, the only thing that I don't understand is why the failed anaconda job with the wrong naming is still there 
Few more next steps remaining to go in downstream:
1. Open a PR that points to the `2023a` branch which cherry-picks this commit after its merging to the main 
2. Add configurations on the CI release repo to build and push the image for the 2023a -> https://github.com/openshift/release/blob/master/ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-2023a.yaml
3.  Sync the downstream main branch with the new commit
4. Add configurations on the CI release repo to build and push the image for the 2023a for the downstream `main` and `2023a` branches config as well https://github.com/openshift/release/tree/master/ci-operator/config/red-hat-data-services/notebooks

**Review by atheo89 — COMMENTED** [review]
/lgtm

**Comment by Xaenalt** [comment]
Ok, package list should be all set

**Comment by Xaenalt** [comment]
One other thing needs fixing on the pkg versions
/hold

File: jupyter/datascience/anaconda-python-3.8/environment.yml
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/environment.yml:343]: lets pin urllib3 under  <1.26.16, as they have removed some function in 2.0 version , which being used in elyra.

File: jupyter/datascience/anaconda-python-3.8/Dockerfile
Comment (harshad16) [inline jupyter/datascience/anaconda-python-3.8/Dockerfile:48]: https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi8-python-3.8/Dockerfile#L43-L44

lets include the runtime images as well.

**Review by harshad16 — CHANGES_REQUESTED** [review]
Thanks for your continuous effort.

With testing, it seems like the install of a separate install of elyra package is causing an issue on the overview page, lets's revert that.

As we would be ready to get this in, let also include runtime image:
https://github.com/opendatahub-io/notebooks/tree/main/jupyter/datascience/ubi8-python-3.8/runtime-images
these would be needed for elyra.
few more suggestion for proper fix.

**Comment by Xaenalt** [comment]
Hmmm, not sure what's going on with e2e tests, seems like it timed out waiting for the notebook?

**Comment by atheo89** [comment]
> Hmmm, not sure what's going on with e2e tests, seems like it timed out waiting for the notebook?

Hey Sean, maybe the problem is with the naming of the notebook into the `.test-%` recipe. 
BTW how do you run the recipes with your customization (deploy, undeploy & test) ? 
aka make deploy8-???-anaconda-python-3.8

**Comment by Xaenalt** [comment]
Since the ubi8 part of the path is hardcoded, I made a separate test target, https://github.com/opendatahub-io/notebooks/pull/28/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R225 and just reused the existing test flow.  My cluster is gone though right now, so I don't have a way to run that part

**Comment by harshad16** [comment]
The issue seems to be the image size, the image is a little bigger, and the pod takes a little more time than timeout to get into a ready state. Testing a few aspects will share.

**Review by harshad16 — COMMENTED** [review]
a suggestion regarding the e2e tests

**Comment by Xaenalt** [comment]
Updated timeout, fingers crossed!

**Comment by harshad16** [comment]
> Updated timeout, fingers crossed!

i dont see any change, can you share where this change was made ?

**Comment by Xaenalt** [comment]
That's weird, it didn't update on the PR

**Comment by Xaenalt** [comment]
Since my cluster is thoroughly dead, only other change that comes to mind is the workdir change that could cause this

**Comment by Xaenalt** [comment]
Just to double check, I ran it with the args from the statefulset, worked, so not sure what's failing e2e

**Comment by zdtsw** [comment]
looks like we need speed up a bit for this PR, since https://github.com/openshift/release/pull/39863 has been merged with testcases which does not have source code available yet.

**Comment by Xaenalt** [comment]
Agreed, it's not clear why e2e is failing on this though

**Comment by atheo89** [comment]
Woohoo 🎉 all the tests passed! 
Shall we merge this now, or you planning to enhance the similar recipes `deploy8/undeploy` and `deploy8-anacoda/undeploy-anacoda` etc into one?

**Comment by Xaenalt** [comment]
I'll add those in later, for now let's just merge this and I'll make the other enhancements in a separate PR

**Pull request merged**
